src/epics/helpers/ChatHandler.js

```diff
@@ -0,0 +1,208 @@
+import { ajax } from 'rxjs/observable/dom/ajax';
+import 'rxjs/add/operator/mergeMap';
+import 'rxjs/add/operator/map';
+import 'rxjs/add/operator/mapTo';
+import 'rxjs/add/operator/catch';
+import 'rxjs/add/operator/do';
+import 'rxjs/add/observable/of';
+import 'rxjs/add/observable/from';
+import 'rxjs/add/observable/throw';
+
+class ChatHandler {
+
+  constructor(wsUrl) {
+    this.wsUrl = wsUrl;
+    this.wsMessageSubject = null;
+    this.wsUserStatusSubject = null;
+  }
+
+  onWsClose = (event) => {
+    console.error('Websocket got close event', event);
+    this.wsMessageSubject.error(new Error('The websocket connection was closed!'));
+  }
+
+  onWsError = (event) => {
+    console.error('Websocket got error event', event);
+    this.wsMessageSubject.error(new Error('Error occured on websocket connection!'));
+  }
+
+  onWsMessage = (event) => {
+    const data = JSON.parse(event.data);
+    if (data && data.ephemeral && data.ephemeral.length && data.ephemeral.length > 0) {
+      Observable.from(data.ephemeral)
+        .mergeMap(ephemeral =>
+          this.createDecryptEphemeralObservable({ ephemeral, mID: this.mID, secretKey: this.secretKey }))
+        .mergeMap(this.resolveIncomingMessageTypeObservable)
+        .catch(error => console.error('An error occured in ChatHandler', error))
+        .subscribe();
+    }
+  }
+
+  resolveIncomingMessageTypeObservable = (message) => {
+    if (message.tags.type === 'chatmessage') {
+
+      const reader = new FileReader();
+      const chatMessageObservable =
+        Observable.fromEvent(reader, 'loadend')
+          .pluck('target', 'result')
+          .map(JSON.parse)
+          .do(contents => {
+            this.wsMessageSubject.next({
+              fromUsername: message.tags.from,
+              maybeSenderId: message.maybeSenderId,
+              message: contents.msg
+            })
+          })
+
+      reader.readAsText(message.fileBlob);
+      return chatMessageObservable;
+    } else if (message.tags.type === 'userstatus') {
+      return Observable.of({
+        username: message.tags.from,
+        status: message.tags.status,
+        created: nowUTC()
+      }).do(status => this.wsUserStatusSubject.next(status));
+    } else {
+      return Observable.throw(new Error('Unrecognized data with type ' + message.tags.type))
+    }
+  }
+
+  createDecryptEphemeralObservable = ({ ephemeral, mID, secretKey }) => {
+    return Observable.create(function (observer) {
+      const binStr = atob(ephemeral);
+      const binStrLength = binStr.length;
+      const array = new Uint8Array(binStrLength);
+
+      for (let i = 0; i < binStrLength; i++) {
+        array[i] = binStr.charCodeAt(i);
+      }
+      const msg = new Blob([array], { type: 'application/octet-stream' });
+      miniLock.crypto.decryptFile(msg,
+        mID,
+        secretKey,
+        (fileBlob, saveName, senderID) => {
+          // NOTE: One tag can have several values. Talk to Steven
```

Comment:
@Spetastian Let's tweak this post-merge

Reply:
That will be my next thing to look into then 😄

Reply:
Perfect